src/collection.js

```diff
@@ -521,8 +504,36 @@ export default class Collection {
    */
   importChanges(syncResultObject, changeObject) {
     return Promise.all(changeObject.changes.map(change => {
-      return this._importChange(change);
+      if (change.deleted) {
+        return Promise.resolve(change);
+      }
+      return this._decodeRecord("remote", change);
     }))
+      .then(decodedChanges => {
+        const changeIds = decodedChanges.map(change => change.id);
+        // XXX: this is not efficient at all.
+        return this.list({}, {includeDeleted: true})
+          .then(res => {
+            const existingRecords = res.data.filter(record => {
+              return changeIds.indexOf(record.id) !== -1;
+            });
```

Comment:
Since we later build a dictionnary based on ids anyway, we don't need this filter this. I had significant performance improvement with removing this filtering.